File: clients/typescript/src/codec.ts
import { encode as msgpackEncode, decode as msgpackDecode } from "@msgpack/msgpack";
import type { Op } from "./protocol.js";
import { PROTOCOL_V1 } from "./protocol.js";

// Wire frame layout (big-endian):
//   u32 payload_len
//   u16 version
//   u16 opcode
//   u32 flags
//   u64 request_id
//   bytes payload[payload_len]
// Total header size: 20 bytes.

export const HEADER_SIZE = 20;

export interface Frame {
  version: number;
  opcode: number;
  flags: number;
  requestId: bigint;
  payload: Uint8Array;
}

/**
 * Serialize a value as msgpack with map/named-field encoding for objects
 * and array encoding for sequences. Matches `rmp_serde::to_vec_named`
 * for protocol frames.
 *
 * NOTE: `@msgpack/msgpack` encodes plain JS objects as maps with string
 * keys and arrays as arrays. That matches `to_vec_named` for the
 * struct-with-named-fields case. User payloads in the Rust client are
 * encoded with `to_vec` (array form for structs); for TS we only
 * serialize protocol frames here, so map encoding is correct.
 */
export function encodeMsgpack(value: unknown): Uint8Array {
  // useBigInt64 ensures bigint values round-trip as msgpack int64.
  return msgpackEncode(value, { useBigInt64: true });
}

/**
 * Deserialize msgpack bytes. Configured to produce bigints for 64-bit
 * integers (matching wire u64 fields like sub_id, offset, delivery_tag.epoch).
 */
export function decodeMsgpack<T>(bytes: Uint8Array): T {
  return msgpackDecode(bytes, { useBigInt64: true }) as T;
}

/**
 * Build a frame for a given opcode and payload value. The payload is
 * msgpack-encoded with map/named-field encoding.
 */
export function buildFrame(op: Op, requestId: bigint, payload: unknown): Frame {
  const bytes =
    payload === null || payload === undefined
      ? new Uint8Array(0)
      : encodeMsgpack(payload);
  return {
    version: PROTOCOL_V1,
    opcode: op,
    flags: 0,
    requestId,
    payload: bytes,
  };
}

/**
 * Encode a frame to its on-wire byte representation.
 */
export function encodeFrame(frame: Frame): Uint8Array {
  const out = new Uint8Array(HEADER_SIZE + frame.payload.byteLength);
  const view = new DataView(out.buffer, out.byteOffset, out.byteLength);
  view.setUint32(0, frame.payload.byteLength, false);
  view.setUint16(4, frame.version, false);
  view.setUint16(6, frame.opcode, false);
  view.setUint32(8, frame.flags, false);
  view.setBigUint64(12, frame.requestId, false);
  out.set(frame.payload, HEADER_SIZE);
  return out;
}

/**
 * Try to decode one frame from the head of a buffer.
 * Returns `null` if not enough bytes are available yet.
 *
 * On success returns the frame and the number of bytes consumed.
 */
export function tryDecodeFrame(
  buf: Uint8Array,
): { frame: Frame; consumed: number } | null {
  if (buf.byteLength < HEADER_SIZE) return null;

  const view = new DataView(buf.buffer, buf.byteOffset, buf.byteLength);
  const payloadLen = view.getUint32(0, false);
  const totalLen = HEADER_SIZE + payloadLen;
  if (buf.byteLength < totalLen) return null;

  const version = view.getUint16(4, false);
  const opcode = view.getUint16(6, false);
  const flags = view.getUint32(8, false);
  const requestId = view.getBigUint64(12, false);

  // Slice owns its own bytes view; copy to detach from the input buffer
  // so the caller can free/reuse `buf` safely.
  const payload = new Uint8Array(payloadLen);
  payload.set(buf.subarray(HEADER_SIZE, totalLen));

  return {
    frame: { version, opcode, flags, requestId, payload },
    consumed: totalLen,
  };
}

/**
 * Decode the typed body of a frame.
 */
export function decodeFrameBody<T>(frame: Frame): T {
  return decodeMsgpack<T>(frame.payload);
}

File: clients/go/codec.go
package fibril

// Frame header and TCP stream framing. The fixed 20-byte header wraps every body
// encoded by the wire codec:
//
//	u32 payload_len
//	u16 version
//	u16 opcode
//	u32 flags
//	u64 request_id
//	bytes payload[payload_len]
//
// Byte-exact with the broker's frame format (crates/protocol/src/v1/frame.rs)
// and every other client.

import "encoding/binary"

// Protocol constants carried in the handshake.
const (
	ProtocolV1       uint16 = 1
	ComplianceString        = "v=1;license=MIT;ai_train=disallowed;policy=AI_POLICY.md"

	frameHeaderSize = 20
)

// Op is a v1 protocol opcode.
type Op uint16

const (
	OpHello    Op = 1
	OpHelloOk  Op = 2
	OpHelloErr Op = 3

	OpAuth    Op = 10
	OpAuthOk  Op = 11
	OpAuthErr Op = 12

	OpPublish        Op = 20
	OpPublishDelayed Op = 21
	OpPublishOk      Op = 25

	OpSubscribe    Op = 30
	OpSubscribeOk  Op = 31
	OpSubscribeErr Op = 32

	OpDeliver           Op = 40
	OpAck               Op = 41
	OpNack              Op = 42
	OpAssignmentChanged Op = 43

	OpPing Op = 50
	OpPong Op = 51

	OpDeclareQueue    Op = 60
	OpDeclareQueueOk  Op = 61
	OpDeclarePlexus   Op = 62
	OpDeclarePlexusOk Op = 63
	OpSubscribeStream Op = 64

	OpReconcileClient Op = 70
	OpReconcileServer Op = 71
	OpReconcileResult Op = 72

	OpTopology   Op = 90
	OpTopologyOk Op = 91
	OpRedirect   Op = 92

	OpTopologyUpdate    Op = 101
	OpTopologyUpdateAck Op = 102
	OpGoingAway         Op = 103

	OpError Op = 255
)

// Frame is a decoded protocol frame: the fixed header fields plus the body.
type Frame struct {
	Version   uint16
	Opcode    Op
	Flags     uint32
	RequestID uint64
	Payload   []byte
}

// buildFrame wraps an already-encoded body in a v1 frame for op.
func buildFrame(op Op, requestID uint64, payload []byte) Frame {
	return Frame{Version: ProtocolV1, Opcode: op, RequestID: requestID, Payload: payload}
}

// encodeFrame serializes a frame to its on-wire byte representation.
func encodeFrame(f Frame) []byte {
	out := make([]byte, frameHeaderSize+len(f.Payload))
	binary.BigEndian.PutUint32(out[0:], uint32(len(f.Payload)))
	binary.BigEndian.PutUint16(out[4:], f.Version)
	binary.BigEndian.PutUint16(out[6:], uint16(f.Opcode))
	binary.BigEndian.PutUint32(out[8:], f.Flags)
	binary.BigEndian.PutUint64(out[12:], f.RequestID)
	copy(out[frameHeaderSize:], f.Payload)
	return out
}

// tryDecodeFrame decodes one frame from the head of buf. It returns the frame
// and the number of bytes consumed, or ok=false when buf does not yet hold a
// full frame (the caller reads more and retries). The payload is copied, so it
// does not alias buf.
func tryDecodeFrame(buf []byte) (frame Frame, consumed int, ok bool) {
	if len(buf) < frameHeaderSize {
		return Frame{}, 0, false
	}
	payloadLen := binary.BigEndian.Uint32(buf[0:])
	total := frameHeaderSize + int(payloadLen)
	if len(buf) < total {
		return Frame{}, 0, false
	}
	f := Frame{
		Version:   binary.BigEndian.Uint16(buf[4:]),
		Opcode:    Op(binary.BigEndian.Uint16(buf[6:])),
		Flags:     binary.BigEndian.Uint32(buf[8:]),
		RequestID: binary.BigEndian.Uint64(buf[12:]),
	}
	if payloadLen > 0 {
		f.Payload = make([]byte, payloadLen)
		copy(f.Payload, buf[frameHeaderSize:total])
	}
	return f, total, true
}
